Windows/DifficultyPredictorWindow.xaml.cs

```diff
@@ -25,144 +26,124 @@ public partial class DifficultyPredictorWindow : Window {
-                foreach (var n in notes) {
-                    var noteTime = 60 / globalBpm * n.beat;
-                    if (windowLower <= noteTime && noteTime <= windowUpper) {
-                        numNotes += 1;
-                    }
-                }
-                densities.Add(numNotes / windowLength);
-                windowLower += step;
-                windowUpper += step;
-            } while (windowUpper < songDuration);
-            return densities;
+
+        private static List<double> GetLocalNoteDensity(List<double> timeSeries, List<double> timeDiff, double windowLength = 2.75) {
+            var cumulativeTime = timeDiff
+                .Select((t, i) => timeDiff.Take(i + 1).Sum())
+                .ToList();
+            var startTimes = timeSeries
+                .Select(t => t - windowLength)
+                .ToList();
+            var startIndices = startTimes
+                .Select(cumulativeTime.BinarySearch)
+                .Select(idx => idx >= 0 ? idx : ~idx)
+                .ToList();
+            var endIndices = timeSeries
+                .Select(cumulativeTime.BinarySearch)
+                .Select(idx => idx >= 0 ? idx : ~idx)
+                .ToList();
+            return startIndices
+                .Zip(endIndices, (start, end) => (end - start) / windowLength)
+                .ToList();
         }
-        // upper 25%
-        private double GetUpperQuartileColumnVariety(List<Note> notes, double songDuration, double globalBpm, double windowLength = 2.75, double step = 0.25) {
-            var variety = new List<double>();
-            var beatsPerWindow = globalBpm / 60 * windowLength;
-            var windowLower = 0.0;
-            var windowUpper = windowLength;
-            while (windowUpper < songDuration) {
-                var localVariety = new List<int>() { 0, 0, 0, 0 };
-                foreach (var n in notes) {
-                    var noteTime = 60 / globalBpm * n.beat;
-                    var noteCol = n.col;
-                    if (windowUpper <= noteTime) {
-                        break;
-                    }
-                    if (windowLower <= noteTime) {
-                        localVariety[noteCol] += 1;
-                    }
-                    var l1Norm = Helper.LpNorm(localVariety, 1);
-                    if (l1Norm > 0) {
-                        // L1-normalise or normalise for the amount of notes
-                        var normLocalVariety = Helper.LpNormalise(localVariety, 1);
-                        // maps with higher column variety will have a distribution closer to [.25, .25, .25, .25]
-                        var score = Helper.LpDistance(normLocalVariety, new List<double>() { 0.25, 0.25, 0.25, 0.25 }, 2);
-                        // higher is better
-                        variety.Add(-1 * score);
-
-                    }
+
+        private static int GetLongestHighDensitySequence(List<double> timeDiffs) {
```

Comment:
As discussed offline, we need to reimplement this function and retrain the model - apart from a bug in the Python code where `series_end` calculates an index of flipped array and doesn't flip it back, what this function calculates is essentially just how often mode value occurs in `timeDiffs`, e.g.:
```
time_diffs_no_zeros=[1, 2, 1, 3, 4, 1, 4]

unique_values=array([1, 2, 3, 4]) indices=array([0, 1, 3, 4], dtype=int64) counts=array([3, 1, 1, 2], dtype=int64)
matrix=array([[1, 1, 1, 0, 0, 0, 0],
       [0, 1, 0, 0, 0, 0, 0],
       [0, 0, 0, 1, 0, 0, 0],
       [0, 0, 0, 0, 1, 1, 0]])
row_sums=array([3, 1, 1, 2]) max_density_index=0
series_start=0 series_end=2
len(get_longest_low_value_sequence(time_diffs_no_zeros)) = 3
```
but the intention was to quantify what's the longest "burst" section in the map.